ground/gcs/src/plugins/pathplanner/flightdatamodel.cpp

```diff
@@ -452,8 +456,118 @@ void FlightDataModel::showErrorDialog(const char* title, const char* message)
+    valPaused = pausing;
+
+    if (!pausing) {
+        fixupValidationErrors();
+    }
+}
+
+void FlightDataModel::fixupValidationErrors()
+{
+    struct FlightDataModel::NED prevNED;
+
+    // Skip validation when there's a download from firmware in process.
+    if (valPaused) { return; }
+
+    for (int i = 0; i < rowCount(); i++) {
+	PathPlanData * row = dataStorage.at(i);
+
+        bool dirty = false;
+
+        struct FlightDataModel::NED thisNED = getNED(row);
+
+        if (i == 0) {
+            row->mode        = Waypoint::MODE_VECTOR;
+            switch (row->mode) {
+		case Waypoint::MODE_CIRCLELEFT:
+		case Waypoint::MODE_CIRCLERIGHT:
+		    row->mode = Waypoint::MODE_VECTOR;
+
+		    showErrorDialog("Waypoint corrected",
+			    "First waypoint may not be the endpoint of an arc");
+                    dirty=true;
+
+		    break;
+		default:
+		    break;
+	    }
+
+        } else {
+	    double distance = sqrt(pow(thisNED.North - prevNED.North, 2) +
+		pow(thisNED.East - prevNED.East, 2));
+
+	    if (distance > 600) {
+		// If the distance is more than 600m, presume it's invalid.
+		// Distances larger than this begin to become problematic
+		// with local tangent plane approximation.
+
+		thisNED = prevNED;
+		setNED(row, thisNED);
+		distance = 0;
+
+		showErrorDialog("Waypoint corrected",
+		        "Over-long leg shortened.");
+
+                dirty=true;
+	    }
+
+	    switch (row->mode) {
+                case Waypoint::MODE_CIRCLELEFT:
+                case Waypoint::MODE_CIRCLERIGHT:
+                    if (row->mode_params < (distance / 2 + 0.1f)) {
+                        row->mode_params = distance / 2 + 0.2f;
+
+                        showErrorDialog("Waypoint corrected",
+                                "Radius of circle increased to minimum");
+
+                        dirty=true;
+                    } 
+
+                    break;
+                case Waypoint::MODE_CIRCLEPOSITIONLEFT:
+                case Waypoint::MODE_CIRCLEPOSITIONRIGHT:
+                    if (row->mode_params < 0.5f) {
+                        row->mode_params = 0.5f;
+
+                        showErrorDialog("Waypoint corrected",
+                                "Radius of circle increased to minimum");
+
+                        dirty=true;
+                    } else if (row->mode_params > 300) {
+                        row->mode_params = 300;
+
+                        showErrorDialog("Waypoint corrected",
+                                "Radius of circle decreased to maximum");
+
+                        dirty=true;
+
+                    }
+
+                    break;
+
+                default:
+                    break;
+            }
+
+            if (dirty) {
+                /* Let anyone listening know we changed it */
+                QModelIndex leftIndex, rightIndex;
+
+                leftIndex = this->index(i, LATPOSITION);
+                rightIndex = this->index(i, LASTCOLUMN-1);
```

Comment:
If we change any of the NED stuff, latitude, longitude, and altitude change too.  So many of the changes change 6 columns.  Some of the changes change the mode or params column.  So I considered it safer to just let UI know that the whole row may have changed.